### PR TITLE
Add guidelines for breaking changes

### PR DESCRIPTION
**Pull request by jpkrohling**
As a result of the discussion involving #4672, we agreed during the SIG Collector meeting that we'd start recommending and enforcing breaking changes to be performed in two steps.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
@dashpole, do you have a good resource about how to do feature flags in the collector? If so, I'll add a link.

File: CONTRIBUTING.md
Comment (codeboten) [inline CONTRIBUTING.md:348]:
```suggestion
### Breaking changes
```

**Review by codeboten — APPROVED** [review]

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:354]: Feature flags do not really help with API changes (only with functionality) since they are evaluated at runtime. Alternative can be "build tags"

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:359]: Not sure if we want to commit to this. Maybe as an "optional"? The helper will be remove anyway next release, so the dependency has to change.

**Comment by bogdandrutu** [comment]
> do you have a good resource about how to do feature flags in the collector? If so, I'll add a link.

https://github.com/open-telemetry/opentelemetry-collector/tree/main/service/featuregate

But as I explained, these are not helping with API changes.

File: CONTRIBUTING.md
Comment (dashpole) [reply CONTRIBUTING.md:354]: Agreed.  I meant it WRT functionality, or configuration.

File: CONTRIBUTING.md
Comment (dmitryax) [inline CONTRIBUTING.md:357]: Maybe we should we explicitly allow moving things across packages and package renames? Otherwise it'll be very difficult to work towards Collector 1.0 given amount of changes we might need to do based on the today's SIG discussions.

**Review by dmitryax — COMMENTED** [review]
Should we explicitly allow package

**Comment by bogdandrutu** [comment]
Question: What is the expected way to change a func signature (breaking change, but assuming it just returns 2 values instead of 1) in the new "world":
1. Create a new `func` and deprecate the old one - downside probably we will not use the best name for the func since that is the old func. Can also be solved by doing this in multiple steps, add a temporary (not ideal name), remove the deprecated one, deprecate the temporary and reuse the good name, remove temporary.
2. Accept some "breaking changes" without deprecation.

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:348]: Question: What is the expected way to change a func signature (breaking change, but assuming it just returns 2 values instead of 1) in the new "world":
1. Create a new `func` and deprecate the old one - downside probably we will not use the best name for the func since that is the old func. Can also be solved by doing this in multiple steps, add a temporary (not ideal name), remove the deprecated one, deprecate the temporary and reuse the good name, remove temporary.
2. Accept some "breaking changes" without deprecation.

deleted the initial comment, and moved it here to allow threads :D

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md:359]: Do we have examples of such helpers?  It's not clear to me what this would look like.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md:357]: If we move types or functions to a new package we should be able to leave behind a forwarding type or function.  If we can't do that because it would introduce circular dependencies then we do it the other way around and have the new type/function reference the prior location until it is removed and the definition moved to the new location.

Comment (dmitryax) [reply CONTRIBUTING.md:357]: Yes, but it still adds a lot of work for us to maintain this approach while this kind of changes is a simple package rename on the client side. I would vote for allowing these changes, let's see what other folks think

Comment (Aneurysm9) [reply CONTRIBUTING.md:357]: pdata is mostly generated code.  We can generate it twice in different packages.  

I think that at this point in the lifecycle of the project it is worthwhile to consider that adding burdens on those proposing making breaking changes is a good thing.  It should perhaps no longer be the case that we make breaking changes in an offhand manner because it is easy to do.

File: CONTRIBUTING.md
Comment (mx-psi) [reply CONTRIBUTING.md:354]: Maybe we should have an "API breaking changes" section and a "End-user facing breaking changes" section?

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:359]: I had the configauth example in mind, where we have a ServerAuthenticator interface, as well as a `NewServerAuthenticator` function that returns an implementation based on the provided options.

https://github.com/open-telemetry/opentelemetry-collector/blob/6b5a3d08a96bfb41a5e121b34f592a1d5c6e0435/config/configauth/default_serverauthenticator.go#L59-L71

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:348]: How would we do it after v1? We'd likely have a new function name that is composed by the ideal name + differences in behavior. Like:

- step 1 (current API): `func GetFoo() Foo`
- step 2 (current+1): needs to return error, then:
- `func MustGetFoo() Foo` (panics on error, is the replacement for `GetFoo() Foo`, which gets deprecated in favor of the new MustGetFoo
- step 3 (current+2): `func GetFoo() (Foo, error)` is added, `func MustGetFoo() Foo` is not touched

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: I added something about it to the doc.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: I added a paragraph about user-facing changes and feature flags, removing it from this paragraph here. Let me know if you think this should be on its own section.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:348]: If we have agreement, I'll add this example to the doc.

File: CONTRIBUTING.md
Comment (mx-psi) [reply CONTRIBUTING.md]: Looks good to me like this!

File: CONTRIBUTING.md
Comment (bogdandrutu) [reply CONTRIBUTING.md:348]: Fine with this. After v1 I will have to survive with the bad name :))

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:348]: What about changing an interface. Adding a func? Do we want to go with optional interface? I don't think there is a way if the interface is used in multiple funcs as argument, because you need to deprecate every one.

Comment (jpkrohling) [reply CONTRIBUTING.md:348]: I don't have the perfect answer, but this question is relevant also for post-v1 and the answer should be the same.

Comment (bogdandrutu) [reply CONTRIBUTING.md:348]: Post v1 I think the willingness to pay any cost is a "given" :))

File: CONTRIBUTING.md
Comment (dashpole) [inline CONTRIBUTING.md:356]: Like the point below ("we create the new name first"), I think we should introduce the replacement code (if it is being replaced) before we deprecate existing functionality?

File: CONTRIBUTING.md
Comment (Aneurysm9) [inline CONTRIBUTING.md:352]:
```suggestion
changelog, preferably with a line instructing users how to move forward.
```

or

```suggestion
changelog, and SHOULD include a line instructing users how to move forward.
```

File: CONTRIBUTING.md
Comment (Aneurysm9) [inline CONTRIBUTING.md:351]:
```suggestion
without a good reason. Hence, when a change is known to cause a breaking change, it MUST be clearly marked in the
```

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md:356]: I very much agree with this, though it is attached to a line about removing indentifiers.  If we're purely removing something without replacement (and with no intent to ever offer a replacement) then there is nothing to have in place before the deprecation notice.

If, however, we're removing something to be replaced with something else, even if it is not a renaming but a change of design, we should definitely have the replacement in place before removing the existing code.  We should also, where possible, change the existing implementation to forward to the new implementation, or re-implement in terms of the new model, before removal.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:357]: Should I add something to the PR to address this?

File: CONTRIBUTING.md
Comment (Aneurysm9) [inline CONTRIBUTING.md:359]: Interfaces (generally) [cannot be added to](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20220114162006-9d54fb35363c/apidiff#readme-interfaces) in a backward compatible manner and I think we should avoid contortions to try to give the appearance that they can.  I also think that we should be [very careful about adding unexported methods](https://github.com/open-telemetry/opentelemetry-collector/issues/4674) to interfaces in an attempt to reserve the ability to add to them later.

The `ServerAuthenticator` helper seems reasonable in that it provides sane default behavior while not being the only way to implement the interface, but some of it is now unnecessary since [nil `StartFunc` and `ShutdownFunc`](https://github.com/open-telemetry/opentelemetry-collector/blob/6b5a3d08a96bfb41a5e121b34f592a1d5c6e0435/component/componenthelper/component.go#L24) have that sane default behavior and an authenticator with default AuthN functionality is useless.  If we were creating this interface today, would we still want this helper or would the existing ways of implementing an `Extension` be sufficient?

I don't think this is the place to discuss the merits of that particular interface, but I do think it might be the case that guidance for how to handle interface compatibility is more complicated and should be discussed and added separately from the rest of this guidance which is more straightforward.

File: CONTRIBUTING.md
Comment (Aneurysm9) [inline CONTRIBUTING.md:364]: A feature flag should probably be added with the deprecation notice, but not yet disabling the feature.  Then the flag default can be set to disable the feature by default, and finally the flag and the feature removed.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:356]: I see where it was ambiguous and have addressed that.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:364]: This is in line with what I had in mind. I changed the text to be clearer.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: I agree with you, but in practical terms, when developing an authenticator, I might not care whether my authenticator is a `ServerAuthenticator` or a `ValidatableServerAuthenticator`. I'd prefer to only provide an `Authenticate` function.

Packages like `confighttp` and `configgrpc` might care about the validation aspect of the component, and `configauth` might provide a default `Validate` that complies with that new interface. IMO, this is the value of providing those helpers: new interfaces can be composed and used in consuming places without having to adapt all the current implementations of the simpler interface.

Comment (Aneurysm9) [reply CONTRIBUTING.md]: In the case of validation I assume the default behavior if no validator is provided is to pass validation.  That could also be achieved by testing for an implementation of a `Validater` interface and only performing validation if it is implemented.  This puts the burden on the consumer, who is in the best position to know whether the new capability is optional (thus taking the interface testing approach) or mandatory (and thus taking an interface composition approach and requiring a `ValidatableServerAuthenticator`).

Comment (Aneurysm9) [reply CONTRIBUTING.md]: I am strongly opposed to this being a `MUST`, as it is in the current revision.  I think if the discussion of interfaces were removed from this PR and handled separately that I could approve this PR, but as it stands I cannot.

File: CONTRIBUTING.md
Comment (dashpole) [reply CONTRIBUTING.md]: I'm still a bit unclear... Which category (removal vs renaming) would general refactoring fall into? For example, if I split `ListAndWatch()` into `List()` and `Watch()`, would that be covered by the removal or renaming category here?  Does "renaming" cover all refactoring for which there is a similar replacement, or are refactors considered removals + new additions?  What would this imply for https://github.com/open-telemetry/opentelemetry-collector/pull/4672#issuecomment-1017804080?

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: Alright, I just removed it from here. We can discuss about it and add a better version in the future.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: Do you think the new version is satisfactory?

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:361]: Is this step necessary? Waiting 2 versions is not enough?

**Comment by bogdandrutu** [comment]
@codeboten lots of changes since your review, I asked you to do another round please.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md:361]: Deprecating features where we impact end users should be where we're most conservative.  I think only two releases between announcement and removal may not be enough time.  The [project guidelines on long-term support](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#long-term-support) are measured in years, not weeks.  We're not there yet, I don't think, but I do think we're beyond the point where a very short notice period is appropriate.  Distributions like ADOT that release once per month could entirely miss the notice period if it is a single release.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md]: I think it's still a little unclear that the feature gate should be added at the same time as the deprecation notice.  That would give users the option to toggle the gate to see how it functions with the feature removed before that becomes the default.  This somewhat extends the time available to detect issues with removing the feature and gives users an option beyond simply filing an issue saying that removing the deprecated behavior might impact them; they can actually test it and know.

File: CONTRIBUTING.md
Comment (bogdandrutu) [reply CONTRIBUTING.md:361]: The current text anyway limits to 2 releases, is just that in the second release you have a gate to enable that back... So what I am asking is if the gate is necessary since we remove it in 2 releases anyway.

Comment (Aneurysm9) [reply CONTRIBUTING.md:361]: We should then `s/the next/a subsequent/` as I don't think a single release window with notice is sufficient.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: > give users the option to toggle the gate to see how it functions with the feature removed before that becomes the default

True, I haven't thought about that. How about this, each step being a version?

1. mark the feature as deprecated, add feature flag with the feature enabled by default
2. change the feature flag to disable it by default
3. remove the flag

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: The rules are the same: MUST mark as deprecated in one version, MAY remove in the next. The difference is that renaming causes the deprecated code to call the new one. So, `ListAndWatch` gets deprecated and calls `List()` and `Watch()`, until it eventually gets removed.

File: CONTRIBUTING.md
Comment (tigrannajaryan) [inline CONTRIBUTING.md:356]: Should we also indicate deprecations in the changelog?

File: CONTRIBUTING.md
Comment (tigrannajaryan) [inline CONTRIBUTING.md:357]: Is this 3 different steps in 3 versions or 2 steps in 2 versions? Wording is not clear.

Also in some rare case it may not possible, e.g. we have A and B and need to swap the names (I can't think of a more realistic use case, but I suspect there may be). I think SHOULD is more appropriate than MUST here.

File: CONTRIBUTING.md
Comment (tigrannajaryan) [inline CONTRIBUTING.md:368]: Log entry may be excessive in some case (if it is in the hot path) and can be against existing guidelines about logging. Let's not overdo it, maybe recommend it as a possibility if it is not in the hot path.

File: CONTRIBUTING.md
Comment (tigrannajaryan) [inline CONTRIBUTING.md:372]: This probably can be generalized for any function signature change, not just return value change or perhaps add one more example to show parameter type/list changes.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:356]: Yes, and it's mentioned earlier in this section:

> Hence, when a change is known to cause a breaking change, it MUST be clearly marked in the
> changelog, and SHOULD include a line instructing users how to move forward

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:357]: Two steps in two versions. I changed the text a bit to try to clarify this.

> Also in some rare case it may not possible, e.g. we have A and B and need to swap the names (I can't think of a more realistic use case, but I suspect there may be). I think SHOULD is more appropriate than MUST here.

How about adding a paragraph about handling exceptions to the rules in general?

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:368]: Agree. I added the word "perhaps" there. Is this sufficient?

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md]: I think Tigran is asking about something slightly different.  That language would apply when a feature is removed.  But before that it should be deprecated.  In many cases there will be a replacement feature added concurrent with the deprecation, and the two should be announced together in the changelog:

> - Added feature X.  This will replace feature Y, which is now deprecated.

If something is being removed with no replacement, though, we should be sure to note its deprecation and pending removal in the changelog.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md]: 3. remove the flag and feature

but, yeah, that's in alignment with what I think will work best.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: I'm sorry, I think I'm still missing something. Can you provide a suggestion to address this? I still see the current text as addressing both situations, requiring changelog entries.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md]: I think it is sufficiently covered.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: Done. I changed the wording a bit, so, I'd appreciate another review of this part.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:348]: Tigran had a comment going the other direction (https://github.com/open-telemetry/opentelemetry-collector/pull/4712#discussion_r789689021): what to do when it's impossible or unfeasible to do it as we described here? I'll add a section about exceptions.

Comment (jpkrohling) [reply CONTRIBUTING.md:348]: I added a new paragraph about the exceptions.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: I added a new paragraph about the exceptions.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md:348]: +1

Rules are meant to be broken, when you know why they exist and why they're the wrong thing to do in a given situation.  I think the callout about consensus among approvers being necessary to make an exception is a good one.  We shouldn't make exceptions with a single approval. 

Perhaps not worth mentioning in this doc, but something we should all keep in mind is that we should take care to ensure that exceptions do not become the norm.  If that is happening then the policy is wrong or we're not being faithful to it and our users.

File: CONTRIBUTING.md
Comment (tigrannajaryan) [reply CONTRIBUTING.md]: Consider scenario:
1. Feature A exists.
1. Feature B is added.
1. After a while Feature B is found to be sufficient to cover use cases where Feature A is used.
1. Feature A is marked deprecated.
1. Feature A is removed.

I think we need to have a changelog about deprecation in step 4. Deprecation _alone_ is noteworthy enough to warrant a changelog entry even if it is not happening simultaneously with some other change.

**Review by tigrannajaryan — APPROVED** [review]
LGTM. This reflects our intent. Any further clarifications can be added via additional PRs.

**Comment by tigrannajaryan** [comment]
@jpkrohling I can merge this once the comments are resolved.

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: So, this?

> - when a feature is being replaced in favor of an existing one, we MUST mark a feature as deprecated in one version, and
>  MAY remove it in the next one

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md]: Can this be marked as resolved, or is there still something I need to clarify in the doc?

File: CONTRIBUTING.md
Comment (jpkrohling) [reply CONTRIBUTING.md:381]: Done

**Comment by jpkrohling** [comment]
PR updated. I think there's only one open comment about features being replaced by existing ones. Not sure there's anything to do in the other open comment.

File: CONTRIBUTING.md
Comment (dashpole) [reply CONTRIBUTING.md]: Can you expand renaming to be clear it covers refactoring as well?
`when renaming or refactoring types, functions, or attributes`...

Comment (jpkrohling) [reply CONTRIBUTING.md]: Got it! Done.

**Comment by jpkrohling** [comment]
> Not sure there's anything to do in the other open comment.

That one was resolved, only one comment is still open but I think it has been addressed already.

File: CONTRIBUTING.md
Comment (Aneurysm9) [reply CONTRIBUTING.md]: `s/the next one/a subsequent version/`

While we are presently trying to move as fast as possible and have a reasonable ability to expect users to adapt to change, these guidelines should outlast this phase, particularly when it comes to features or behavior for end users that may not be compilation breaking.  We should be prepared to have deprecation periods that last not days or weeks, but many months spanning several releases.

Comment (jpkrohling) [reply CONTRIBUTING.md]: My bad, I used "a subsequent one" elsewhere and should have used it here as well. Fixed now.

**Review by Aneurysm9 — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

File: CONTRIBUTING.md
Comment (bogdandrutu) [inline CONTRIBUTING.md:405]: should we define what consensus means? half + 1? You are defining something here that is not well explained.

Comment (jpkrohling) [reply CONTRIBUTING.md:405]: I understand "consensus" as unanimity, but I'm not a native speaker. My intention was that we, as engineers, should be able to find a solution that satisfies everyone, even if it means making compromises along the way.

Comment (Aneurysm9) [reply CONTRIBUTING.md:405]: I don't think consensus needs to mean unanimity, but generally at least that there are not strong/formal objections.  I'm a fan of the [w3c consensus model](https://www.w3.org/2021/Process-20211102/#Consensus).

I think that we will not always be able to satisfy everyone, but we must ensure that we have exhausted reasonable attempts to do so and have policies and procedures for dealing with formal objections.  We should generally strive to minimize opposition and look for consensus options that have the most support and least resistance.

Comment (jpkrohling) [reply CONTRIBUTING.md:405]: Do we need to specify in this doc what consensus means to us? Do we know what it means to us?

Comment (Aneurysm9) [reply CONTRIBUTING.md:405]: I would prefer that consensus be defined at a higher layer than contributing docs.  It is a concept that should apply to most things we do and applies not only to this SIG.  Perhaps we need to define for ourselves how to apply the general concept in particular situations, but that should probably be done in reference to a globally used definition.

Comment (bogdandrutu) [reply CONTRIBUTING.md:405]: Without defining what it means, how can we apply this?

Comment (jpkrohling) [reply CONTRIBUTING.md:405]: We might be overthinking this, but here's a proposal: 

>  A reason for requesting the exception MUST be given in the pull request. Until unanimity
> is obtained, approvers and maintainers are encouraged to discuss the issue at hand. If a consensus (unanimity) cannot be
> obtained, the maintainers are then tasked in getting a decision using its regular means (voting, TC help, ...).

Comment (jpkrohling) [reply CONTRIBUTING.md:405]: @bogdandrutu, is this proposal acceptable?

Comment (bogdandrutu) [reply CONTRIBUTING.md:405]: Ok, sounds reasonable as long as the approvers are reasonable :)

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
